### PR TITLE
WIP: Scripts and configs to run k8s cluster on GCE using CoreOS as VM image

### PR DESCRIPTION
**Pull request by Amey-D**
This is a work-in-progress, because all e2e tests do not pass.

Tested:

The end-to-end test runs to completion with the following setup:

export KUBERNETES_PROVIDER="gce"
export KUBE_GCE_IMAGE="coreos-stable-557-2-0-v20150210"
export KUBE_GCE_IMAGE_PROJECT=coreos-cloud
hack/e2e-test.sh

e2e test summary:
...
Ran 26 of 26 Specs in 3892.090 seconds

FAIL! -- 17 Passed | 9 Failed | 0 Pending | 0 Skipped F0305
23:48:33.498276    6334 driver.go:87] At least one test failed

2015/03/05 23:48:33 Error running Ginkgo tests: exit status 255

### CODE REVIEW

**Comment by zmerlynn** [comment]
Can you summarize what's failing right now? 9 is a fair amout

**Comment by zmerlynn** [comment]
(Self-assigning because I've been working with @dyema internally)

cc @kelseyhightower: BTW, this is what I was talking to you about at the meet-up. We're working on getting it out in the open and passing e2es now, but I'm sure @dyema would love any input on the approach from Core.

**Comment by dyema** [comment]
Summarizing 9 Failures:

[Fail] Shell [It] tests that guestbook.sh passes 

/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/shell.go:67

[Fail] Shell [It] tests that liveness.sh passes 

/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/shell.go:67

[Fail] Cadvisor [It] should be healthy on every node. 

/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/cadvisor.go:75

[Fail] Services [It] should provide DNS for the cluster 

/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:153

[Fail] PD [It] should schedule a pod w/ a readonly PD on two hosts, then remove both. 

/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/pd.go:121

[Fail] Shell [It] tests that pd.sh passes 

/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/shell.go:67

[Fail] PD [It] should schedule a pod w/ a RW PD, remove it, then schedule it on another host 

/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/pd.go:80

[Fail] MasterCerts [It] should have all expected certs on the master 

/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/certs.go:49

[Fail] Shell [It] tests that services.sh passes 

/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/shell.go:67

Ran 26 of 26 Specs in 3892.090 seconds

FAIL! -- 17 Passed | 9 Failed | 0 Pending | 0 Skipped F0305 23:48:33.498276    6334 driver.go:87] At least one test failed

2015/03/05 23:48:33 Error running Ginkgo tests: exit status 255

**Comment by dyema** [comment]
Also, thanks to @rlarocque2 for initial prototype and to @andyzheng0831 for helping out with debugging.

**Comment by dyema** [comment]
@zmerlynn I suspect the monitoring, PD and DNS tests are failing because the features are not properly configured in the scripts (known issues).  Not sure about the rest, will take another look.

**Comment by dchen1107** [comment]
FYI: cc/ @thockin @brendandburns @bgrant0607 

Please note that this is image is similar to containervm image except it is based on latest coreos distro, and containervm image is based on debian.

**Comment by thockin** [comment]
Does it include socat and nsenter and git and iptables binaries?   How
about the GCE scripts like /usr/share/google/safe_format_and_mount?

**Comment by filbranden** [comment]
@filbranden who wants to review this at least for bash :-)

**Comment by dyema** [comment]
@thockin CoreOS stock image seems to include nsenter, iptables, git (not sure why k8s requires git) and safe_format_and_mount.  Didn't find socat (in CoreOS stable (557.2.0) at least).

File: cluster/gce/coreos/README.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+### Notes
+The script invokes two "gcloud compute ssh" commands per VM instance -- to copy data over and to reboot the machine.  For convenient cluster setup, it is recommended that "gcloud compute ssh" works without password from the machine where the users intend to run "kube-up.sh".  Please refer to ["gcloud compute ssh](https://cloud.google.com/sdk/gcloud/reference/compute/ssh) and ["Connecting to an instance using ssh"](https://cloud.google.com/compute/docs/instances#sshing) for more details.
+
+### Commands
+Please refer to [kubernetes docs](docs/getting-started-guides/binary_release.md) for instructions on checking out and building from sources.  It is assumed that the user can run a kubernetes cluster on GCE using "containervm" image for the VMs.
+
+```bash
+export KUBERNETES_PROVIDER="gce"
+export KUBE_GCE_IMAGE="coreos-stable-557-2-0-v20150210"
+export KUBE_GCE_IMAGE_PROJECT=coreos-cloud
```
Comment (zmerlynn) [inline cluster/gce/coreos/README.md]: Is there any way we could maybe get just the last bit (`coreos-cloud`) to imply the first bit (`coreos-stable-557-2-0-v20150210`)? It would be convenient to bake a default image in, even if it's slightly annoying.

File: cluster/gce/coreos/config-default.sh [hunk 2]
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash
+
+# Copyright 2014 Google Inc. All rights reserved.
```
Comment (zmerlynn) [inline cluster/gce/coreos/config-default.sh]: Nit: 2015

File: cluster/gce/coreos/config-test.sh [hunk 3]
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash
+
+# Copyright 2014 Google Inc. All rights reserved.
```
Comment (zmerlynn) [inline cluster/gce/coreos/config-test.sh]: Nit: 2015

File: cluster/gce/coreos/container_bridge.sh [hunk 4]
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+
+set -e
```
Comment (zmerlynn) [inline cluster/gce/coreos/container_bridge.sh]: errexit for consistency with codebase

File: cluster/gce/coreos/container_bridge.sh [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+set -o pipefail
```
Comment (zmerlynn) [inline cluster/gce/coreos/container_bridge.sh]: most of our scripts also `set -o nounset` for extra persnitickity

**Comment by thockin** [comment]
We have a "git volume" which execs git.  We need socat for port forwarding,

File: cluster/gce/coreos/master.yaml [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        After=network-online.target
+        Requires=kubernetes-node-params.service
+        After=kubernetes-node-params.service
+
+        [Service]
+        Type=oneshot
+        RemainAfterExit=yes
+        EnvironmentFile=/etc/kubernetes-node-params
+        ExecStartPre=/usr/bin/mkdir -p /opt/kubernetes/pkg
+        ExecStartPre=/usr/bin/curl --location --create-dirs --output /opt/kubernetes/pkg/kubernetes-server-linux-amd64.tar.gz ${SERVER_BINARY_TAR_URL}
```
Comment (zmerlynn) [inline cluster/gce/coreos/master.yaml]: Borrowing from `download-or-bust`, I would add something like `--connect-timeout 20 --retry 6 --retry-delay 10` to any of the `curl` commands. It's really bad if we can't get the software.

File: cluster/gce/coreos/util.sh [hunk 7]
```diff
@@ -0,0 +1,1035 @@
+#!/bin/bash
+
+# Copyright 2014 Google Inc. All rights reserved.
```
Comment (zmerlynn) [inline cluster/gce/coreos/util.sh]: Nit: 2015

**Comment by rlarocque2** [comment]
IIRC, CoreOS ships with netcat.  dyema@, could you confirm?  If it does exist, could you use it instead of socat?

**Comment by thockin** [comment]
I am not sure. netcat (depending on version) has weird issues wrt stdin.
The current port forwarding code is written to socat.  Maybe possible.

**Comment by erictune** [comment]
Is this intended to be a distro that developers would use?  If so, I'm not sure I like that developers now have to learn and modify both salt and cloudconfig.  Don't get me cloudconfig is pretty awesome, but two ways to do things isn't.

On the other hand, if this is just for people who want to setup a cluster from a binary release, then I don't think cluster/ is the right place for this.

(where I am coming from:  #5141)

**Comment by zmerlynn** [comment]
@erictune: That was a comment I made before you sent it out more broadly. I don't think we can be stuck with just Salt as the only dev path - we're proposing this as an in-tree, continuously integrated CoreOS provider. @dyema still has a fair amount of work to do before we can get it fully baked, but we wanted to get eyes on the PR and concept.

**Comment by piosz** [comment]
Is this PR active? If not please close it according to https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/devel/pull-requests.md

**Comment by dyema** [comment]
I had a chat with zmerlynn@ et al internally, and this is not the highest priority at the moment.  Closing this.
